core/src/main/java/com/sap/ai/sdk/core/AiCoreServiceStub.java

```diff
@@ -0,0 +1,98 @@
+package com.sap.ai.sdk.core;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.google.common.collect.Iterables;
+import com.sap.cloud.sdk.cloudplatform.connectivity.ApacheHttpClient5Accessor;
+import com.sap.cloud.sdk.cloudplatform.connectivity.Destination;
+import com.sap.cloud.sdk.services.openapi.apiclient.ApiClient;
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+import org.springframework.http.client.BufferingClientHttpRequestFactory;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder;
+import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
+import org.springframework.web.client.RestTemplate;
+
+/** Container for an API client and destination. */
+@FunctionalInterface
+public interface AiCoreServiceStub {
+  /**
+   * Get the destination.
+   *
+   * @return the destination
+   */
+  @Nonnull
+  Destination destination();
+
+  /**
+   * Get the API client.
+   *
+   * @return the API client
+   */
+  @Nonnull
+  default ApiClient client() {
+    return client(ClientOptions.SERIALIZE_WITHOUT_NULL_VALUES);
+  }
+
+  /**
+   * Get the API client with options.
+   *
+   * @param options the options
+   * @return the API client
+   */
+  @Nonnull
+  default ApiClient client(@Nonnull final ClientOptions options) {
+    final Destination destination = destination();
+    return options.getInitializer().apply(destination);
+  }
+
+  /** Options for the API client. */
+  interface ClientOptions {
+
+    /** Serialize with null values. */
+    ClientOptions SERIALIZE_WITH_NULL_VALUES = () -> ApiClient::new;
+
+    /** Serialize without null values. */
+    ClientOptions SERIALIZE_WITHOUT_NULL_VALUES = () -> ClientOptions::withoutNull;
```

Comment:
Why? Do you see any other usecase outside of those 2?

Reply:
Everything can be achieved by creating your own implementation of `AiCoreServiceStub` (or extending `AiCoreService`) and changing the implementation of `client()`.
This would cut the `AiCoreServiceStub` in half.

Reply:
In general I would agree with you, users providing their implementation of `client()` would be preferred over additional/optional `client(options)`.

Unfortunately, due to the immutable nature of the classes, all accessors, like `withResourceGroup(String)` will instantiate a new object, that escapes the custom/overridden methods.

If the user still wanted to override while having all features, the code would look like the following, given the current state:

<details><summary>(1) base-case, assuming productive code does not make use of deployment API</summary>

```java
class CustomAiCoreServoce extends AiCoreService {
    @Nonnull
    @Override
    public ApiClient client() {
        return new ApiClient(destination()); // re-activates null-serialization
    }
}
```

</details>
<details><summary>(2) extended case, user knows deployment id</summary>

```java
final AiCoreService base = new AiCoreService();
class CustomAiCoreDeployment extends AiCoreDeployment {
    public CustomAiCoreDeployment() {
        super(res -> "my-deployment-id", base::destination);
    }

    @Nonnull
    @Override
    public ApiClient client() {
        return new ApiClient(destination()); // re-activates null-serialization
    }
}
```

</details>
<details><summary>(3) extended case, user knows deployment is identified via scenario</summary>

```java
final AiCoreService base = new AiCoreService();
class OtherAiCoreDeployment extends AiCoreDeployment {
    public OtherAiCoreDeployment() {
        super(res -> getDeploymentId(base.client(), res), base::destination);
    }

    @Nonnull
    @Override
    public ApiClient client() {
        return new ApiClient(destination()); // re-activates null-serialization
    }

    @Nonnull
    static String getDeploymentId(@Nonnull final ApiClient client, @Nonnull final String res) {
        final Predicate<AiDeployment> p = deployment -> "orchestration".equals(deployment.getScenarioId());
        return AiCoreDeployment.getDeploymentId(client, res, p);
    }
}
```

</details>

While the base-case looks fine for me, the extended cases are messy.

Also I'm not sure how much effort we should put further into **the underlying issues** with `ApiClient` from _SAP Cloud SDK_, since we are planning to explore alternatives soon. According to our backlog.